keps/sig-api-machinery/20190820-master-service-of-type-externalname.md

```diff
@@ -0,0 +1,302 @@
+
+It might be more suitable to advertise kube-apiserver's endpoint as FQDN, such as `api.my-cluster.example.com` for disaster recovery / prevention - in case of accidental deletion of said load-balancer and/or it's associated IP, the only thing required for recovery is to reconfigure the NS records to point to the newly created load-balancer.
+
+Historically there are several issues related to this problem:
+
+- https://github.com/kubernetes/kubernetes/issues/34043
+- https://github.com/kubernetes/kubernetes/issues/47588
+- https://github.com/coreos/tectonic-installer/issues/384
+
+### Goals
+
+- Make kube-servers exposed and advertised as FQDN, first class citizen in Kubernetes.
+- The master service reconciler in kube-apiserver should support services of type `ExternalName`.
+- Support in kubelet for injecting the required environment variables `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` for [in-cluster discovery](https://github.com/kubernetes/client-go/blob/c8dc69f8a8bf8d8640493ce26688b26c7bfde8e6/rest/config.go#L399-L411) for master kubernetes service of type `ExternalName`.
+
+### Non-Goals
+
+- Environment variable support in kubelet for services of type `ExternalName` other than the master service. Related [comment](https://github.com/kubernetes/kubernetes/issues/60535#issuecomment-490757630).
+- In-cluster API server detection/auto-configuration other than environment variables named `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT`.
+- Network connectivity from Nodes/Pods to the kube-server is out of scope and is assumed to be properly configured and working. Pods using the hostnetwork or CNI network are able to initiate connections to kube-apiserver.
+
+## Proposal
+
+In order to support this feature in Kubernetes the following changes are proposed:
+
+### Kube-apiserver
+
+The master service reconciler in kube-apiserver is extended to support services of type `ExternalName`.
+
+Example:
+
+Given a kube-apiserver advertised with host `api.127.0.0.1.nip.io` and port `443`, it will reconcile the master service with the following content:
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    component: apiserver
+    provider: kubernetes
+  name: kubernetes
+  namespace: default
+spec:
+  clusterIP: ""
+  type: ExternalName
+  externalName: api.127.0.0.1.nip.io
+  ports:
+  - name: https
+    port: 443
+    protocol: TCP
+    targetPort: 443
+```
+
+It'll also delete the corresponding `Endpoints` called `kubernetes` from the `default` namespace as it's not needed for `Services` of type `ExternalName`.
```

Comment:
just asking because I don't know exactly how this service work: 
the `externalName` creates a CNAME record but if you delete the endpoints, how are the A/AAAA records generated?

Reply:
`KubeDNS` / `CoreDNS` generate `CNAME` record from the service's name and this record points to the `externalName` field of the `Service`. No endpoints are used in this case.

Reply:
yeah, but the who resolves `externalName` to an IP address?

```
kubernetes CNAME apiserver.local
apiserver.local A 10.96.0.1
```

Reply:
Node's DNS resolver - `KubeDNS`/`CoreDNS` will query for `apiserver.local` and it'll get it from that resolver.

Reply:
this is a very strong dependency on the node resolver, don't you think so?

Reply:
Yes it does. Unfortunately, a DNS resolution on the Nodes is more or less a _must_ at this point - image registries are hosted and referenced in Pods. For example, by default, `kubelet` uses `k8s.gcr.io/pause` as a pause image. 

Running a cluster without a a DNS resolution on the Node is possible, but at this moment it would require lots of tinkering with different configurations.

Reply:
Sorry for keep asking, but this is my doubts, independently of the "whole" DNS resolution, and continuing with current example of `apiserver.local`, in some place there has to be an entry that maps `apiserver.local` to an IP address, with this approach it seems it will be in a external DNS server or `/etc/hosts`.

Reply:
seems this comment answer my question, managing it manually out of kubernetes
https://github.com/kubernetes/enhancements/pull/1216#discussion_r370903033